Better file dialogue

ShlomiRex opened the pull request:
This branch is for making file dialogue like file explorer, with Last modified and Size of files and folders and user can sort by those columns, support for diffirent column placement (last modified can be first column, etc..)

**Source branch:**
master

![Peek 2020-02-07 16-53](https://user-images.githubusercontent.com/17744127/74293083-696aab00-4d42-11ea-89cd-4478aef93aaf.gif)


**Bugs to fix:**
* ~~When selecting an already selected row, it deselects it. Note:  [Issue #36047](https://github.com/godotengine/godot/issues/36047)~~
* When going up a directory or into directory I get this: (It still works, just throws error)

> ERROR: deselect_all: Condition "item == prev_item" is true.
   At: scene/gui/tree.cpp:3210.

**Features to implement:**
* ~~Instead of tree title being just a string, turn it to button so when it is clicked, file dialogue will sort the entries by that column.~~ Currently it is very complex maybe in the future
* ~~Add more rich Date options, like date formats and such. Instead of using _OS singleton to get date you can use Date Object.~~ People don't like this feature, also it should be other pull request, unrelated to UI
* ~~Columns's width should change like when using [SliderContainer](https://docs.godotengine.org/en/3.1/tutorials/gui/gui_containers.html#split-container)~~ Like the first, it is very complex, maybe in the future it will be available

Issues closed: 
* Issue #36047 - This pull request uses #36094 pull request

pouleyKetchoupp committed:
Option in RichTextLabel for height to fit content

(cherry picked from commit ad8081216c73b8a0ace555eeb96c1d4fbcb3e50e)

Calinou committed:
Add visual feedback when hovering layer checkboxes in the Inspector

This also changes how checkboxes are selected, which makes it possible
to click in the small area between two checkboxes and
still toggle a value successfully (which is arguably less frustrating).

(cherry picked from commit bbc435624f6569660eb12203b7c32c87b4027ecf)

madmiraal committed:
Check for motion in cast_motion() before doing Bullet convexSweepTest().

Also ensure that default closest_safe and closest_unsafe values are
defined in cast_motion() and before cast_motion() is called.

Calinou committed:
Improve `CONTRIBUTING.md` and update it to follow recent changes

- Improve the bug reporting guidelines for readability.
  - Make some guidelines more strictly applied in the interest of
    bug report quality.
- Mention that feature proposals should now be opened on the
  Godot Proposals repository.
- Update the Git commit message guide to be easier to follow.
- Fix warnings reported by markdownlint.

(cherry picked from commit 7b5c502ba5c69c68003441a425981a69ca4876c2)

Calinou commented:
By the way, you can use `some_string.pad_zeros(2)` to pad dates.

YeldhamDev commented:
It would be nice to have those enhancements in the `EditorFileDialog` as well.

ShlomiRex commented:
> Thanks for working on this, it will be much appreciated slightly_smiling_face
> 
> I noticed dates on the screenshot aren't padded; you can use `some_string.pad_zeros(2)` to pad dates/times. Also, it's probably a [better idea](https://xkcd.com/1179/) to default to an ISO 8601-like format such as `YYYY-MM-DD HH:MM` if we can't detect the user's date format easily.

Thanks ill try to impliment it.
Is there a way to have a Tree with the TreeItem (rows) being other Controls ?
Like Tree that all the children are Buttons?
Moreover I fixed the de-select problem

Lastly, what do you think about having adjustable column's size? Like Split container

Calinou commented:
> Lastly, what do you think about having adjustable column's size? Like Split container

I guess this would have to be implemented in Tree, but it's a pretty complex class already.

ShlomiRex commented:
Pull request is ready for merge

Zylann commented on core/os/file_access.cpp:
```diff
@@ -501,10 +501,59 @@ uint64_t FileAccess::get_modified_time(const String &p_file) {
+}
+
+FileAccess::FileSize FileAccess::get_size_simplified(const String &p_file) {
+	Dictionary dic = get_sizes(p_file);
```
Allocating a dictionary containing several multiples of of the same original size only to return it as a one size sounds a bit wasteful.

dreamsComeTrue committed:
Allow scroll_to_line when scroll_active is 'false'

Fix #36134

(cherry picked from commit 8f11a91917bc85976390477712dc4202ab383fe7)

KoBeWi committed:
Add unique icon to Polygon2D

(cherry picked from commit 225622e5e3258e4b18c3d5c34a5050fe3fa82624)

nodj committed:
2D Editor: modified zoom increment to the twelveth root of two

- properly visit power of 2 factors (50%, 100%, 200%...)
- index based zoom values to prevent floating point issues
- Fix 2d editor not able to reach min and max zoom values

(cherry picked from commit fea6ca20c9b34f077396210dc713f79f0918be5f)

ThakeeNathees committed:
colorpicker margin bug fixed

(cherry picked from commit 9d224ac42920728031ac389dfc11f7fa9e190176)

Xrayez committed:
SCons: Dump construction environment to a file

A new `methods.dump(env)` is added to dump the construction environment
used by SCons to build Godot to a `.scons_env.json`. The file can be used
for debugging purposes and any external tool.

(cherry picked from commit 42bee75e86e81fcd9b51b2e960d13b89b18ad4bf)

akien-mga committed:
SCons: Format buildsystem files with psf/black

Configured for a max line length of 120 characters.

psf/black is very opinionated and purposely doesn't leave much room for
configuration. The output is mostly OK so that should be fine for us,
but some things worth noting:

- Manually wrapped strings will be reflowed, so by using a line length
  of 120 for the sake of preserving readability for our long command
  calls, it also means that some manually wrapped strings are back on
  the same line and should be manually merged again.

- Code generators using string concatenation extensively look awful,
  since black puts each operand on a single line. We need to refactor
  these generators to use more pythonic string formatting, for which
  many options are available (`%`, `format` or f-strings).

- CI checks and a pre-commit hook will be added to ensure that future
  buildsystem changes are well-formatted.

(cherry picked from commit cd4e46ee65dab6baa6a143bf3b3f64244be36712)

akien-mga committed:
Hooks: Add pre-commit hook for psf/black formatting

(cherry picked from commit 164826a39bca2fb7b7277752cbc1df8833ce0f1a)

akien-mga committed:
Hooks: Use pygmentize if available to visualize diff

(cherry picked from commit 4d52761da6f15ee0374a4cac958cc7cd12507adc)

akien-mga committed:
Travis: Add static check for Python black formatting

Also install and use pygmentize to visualize clang-format and black
diffs.

(cherry picked from commit 3644036fd3d8678ac44695cc49ae63c4aaeb1b97)

Xrayez committed:
Moved to methods.py as a `show_progress` method.

Some required changes are made:
- locally imported SCons-specific packages within the method;
- `global` variables converted to `nonlocal` (used in nested functions).

(cherry picked from commit d753a7630ae0cd20ce7e769346b0329f28ab8de9)

pouleyKetchoupp committed:
Pre-commit hook instructions on Windows

(cherry picked from commit 7b6e664178ce0ff559a7a5033606a089881f21b4)

azagaya committed:
Fixing wrong blending rect methods

Using Color.blend function instead of custom code

Fixed clang_format

Removed unnecessary help

(cherry picked from commit b211a86ebe241599f95186f9db28be2061a19962)

bruvzg committed:
Git Hooks: Add support for GUI git clients.
[ci skip]

(cherry picked from commit fdad0e3bd1e2f8dcfcdeae7c2617eda5ac24324e)

Xrayez committed:
Bind Shape2D draw method

(cherry picked from commit 430d1fd79544906e9b55092441f2f446948cbf2e)

ThakeeNathees committed:
GDScript debugger incorrect error line fixed

if the first line of an else or an elif throws a runtime error the
debugger shows incorrect line number.

(cherry picked from commit 9325671faa9d8407996effd4366f47733a1ec21b)

bruvzg committed:
Git Hooks: Suppress "which" error messages.

(cherry picked from commit 05f041127d3811fe44403d72784ac7553f6681c3)

RandomShaper committed:
Fix potential crash when listing leaked objects

Note:
Casting to the C++ classes and calling the methods there would work as well,
but would require including he header files for the specific object types handled
here, which wouldn't be OK either.

rsubtil committed:
Allow Godot to kill it's own PID

(cherry picked from commit 75ce45440a9a2cd6363b5f04205d201dba6dec5d)

akien-mga committed:
xatlas: Sync with upstream 470576d

(cherry picked from commit 59780fd046f4bcecbf892a47e47e2f22068b5d7c)

Calinou committed:
Tweak the import compression property hint for clarity

(cherry picked from commit 1439447071afa35241f2b9453132fd2282aa3eab)

KoBeWi committed:
Link exposed set/getters in property descriptions

(cherry picked from commit ea07acb877b9c888c281f7af65b0a22f147e8426)

lawnjelly committed:
Fix overflow condition with QueryPerformanceCounter

The previous code for OS_Windows::get_ticks_usec() multiplied the tick count by 1000000 before dividing by ticks_per_second. The ticks is counted in a 64 bit integer and is susceptible to overflow when a machine has been running for a long period of time (days) with a high frequency timer.

This PR separates the overall calculation into one for seconds and one for the remainder, removing the possibility of overflow due to the multiplier.

(cherry picked from commit db9fa8816056b367a5d97f57c2498241aaef974a)

pouleyKetchoupp committed:
Fix dialog spam when inspecting MeshInstance from model file

Avoid load_scene for built-in resources to make sure we don't
open a scene tab and prompt for model file editing.

Load scene as regular resource instead and store the reference to keep
the dependency until the remote inspector cache is cleared.

(cherry picked from commit 01802074a01c41d8672664751f80b80069796f12)

abustin committed:
Respect 'mesh compression' editor import option in Assimp (ie. FBX) and glTF importers

mbrlabs committed:
Perform a clean Gradle build if android plugins changed

Fixes #38986

(cherry picked from commit abd66c1bb09b85199ed0e513d63fd6b827e4b21b)

Calinou committed:
Improve the text appearance in the script editor warnings panel

- Make the Ignore button's position identical across all warnings
  by moving it to the left.
- Change the Ignore button's text and color to make it more obvious
  that it can be clicked.
- Use the editor font instead of the default project font to match
  the rest of the editor.

(cherry picked from commit 0c0ee427d27f09f8d01cb46bf59e144c1ba30c5e)

Commit:
Fix ProjectSettings.save_custom documentation

(cherry picked from commit 728c0c1e14d13f7aa1f48e3b110bee9b6fcc6d77)

Grosskopf commented on scene/gui/file_dialog.cpp:
```diff
@@ -702,7 +788,7 @@ void FileDialog::set_mode(Mode p_mode) {
 		tree->set_select_mode(Tree::SELECT_MULTI);
 	} else {
-		tree->set_select_mode(Tree::SELECT_SINGLE);
+		tree->set_select_mode(Tree::SELECT_ROW);
```
Due to this Line, When running tree->deselect_all() in line 346 the item->deselect(selected_col) in tree.cpp line 3210 fails to deselect since it is not among the Modes mentioned in tree.cpp line 3190, i am not sure why that is, still figuring out why the if clause in tree.cpp line 3190 makes sense, but that is how you end up with the `ERROR: deselect_all: Condition "item == prev_item" is true.`

Grosskopf replied:
i checked, returning it to SELECT_SINGLE fixes the issue, not sure why SELECT_ROW was needed, you are still selecting a single file but maybe it has some other effects, i have no idea :D

ShlomiRex replied:
> i checked, returning it to SELECT_SINGLE fixes the issue, not sure why SELECT_ROW was needed, you are still selecting a single file but maybe it has some other effects, i have no idea :D
Because if you do this is what you get
![Screenshot from 2020-06-18 00-26-29](https://user-images.githubusercontent.com/17744127/84952267-6ae61e80-b0fa-11ea-8593-6b92864e351d.png)

Grosskopf commented:
yay, i found where the issue is coming from, i'm no expert in the tree.cpp, but maybe @ShlomiRex you know what to adjust from my comment in the file_dialog.cpp ^^

mrushyendra committed:
Document usage requirement for Viewport to sample ViewportTexture as HDR

Modified class documentation to specify that `USAGE_3D` or
`USAGE_3D_NO_EFFECTS` is required in order to use HDR in a ViewPort.

Fixes #38702

swarnimarun committed:
Fix crashing of VisualScript due to...

Attempting to move the function node to another function whose data connection is a dependency of the node the specific node being moved to a different function during changes to sequence connections.
By skipping, if the from_node is a function_node during the data connection dependencies scan.
Should fix #37991

(cherry picked from commit 5c48631509751191afc5a83e4c48075a90a38150)

akien-mga committed:
Improve documentation of ProjectSettings::setup

Subsequent PRs adding and fixing support for bundled PCKs
did not update the documentation /o\

(cherry picked from commit 0183bfbb0f8481cbb7886b48c8d55b788b92e253)

madmiraal committed:
Update Windows GUID to SDL uid conversion.

(cherry picked from commit a833db80d907c57630a64e449f3667489e8a0005)

m4gr3d committed:
Stop ignoring hidden files and directories in the `assets` directory

(cherry picked from commit bf76d2afcf6d36fa26acb12d52031130e6ce012b)

mrushyendra committed:
Update Mono C# warning to reflect latest list of platforms supported

Simple fix for #38627. iOS (#20268) and HTML5 (#20270) removed from list of exceptions
for platforms supported in warning message.

(cherry picked from commit 3d03be7a56769b6188c6ab8df6ca0c41877bc61c)

akien-mga committed:
Add shortcut for Pan Mode (G)

And change TileMap Bucket Fill shortcut to (B).

Fixes #8582.

(cherry picked from commit 28b33718b00581d28fd56c83761077eef19893be)

Commit:
Do not rely on the existence of `config.py` while detecting modules

The existence of `SCsub` is checked instead. This file is required for
all modules, and prevents the build system to leave modules without
`config.py` undetected, leading to silently ignoring the module during
compilation.

(cherry picked from commit 12389b37363db541517c1c722e2e82294e92e251)

bruvzg committed:
Ignore invalid tablet driver name, when non are available.

(cherry picked from commit ef1a305e50331611db628a25ca777ffe824d37f3)

vnen committed:
GDScript: Fix assert message when no custom message is set

(cherry picked from commit f29a2e2606317e0964ffed287187afe6bf115626)

mrushyendra committed:
Expose `get_char_size()` from Font instead of BitmapFont

`get_char_size()` is a public virtual function defined in the `Font`
class. Implementations exist for both `BitmapFont` and `Dynamic Font`.
However, it was only exposed to the GDScript API through the Bitmap
Font, and not for Dynamic Font.

This commit exposes the function through `Font` instead.

Fixes #23967

(cherry picked from commit a4413710f9418bd5a5ab83ac3ed70b1329cb2492)

YeldhamDev committed:
Add fallback icons and make custom ones appear in the recent/favorites list in the "Create New" dialog

(cherry picked from commit 84d7492b2d9e2c2b8de84327e98783336cbbf35d)

Calinou committed:
makerst: Print status messages to make the CI output clearer

This also removes an unused method.

(cherry picked from commit 9065b7ed4e72d107362a55a1d2b3ded20ae3a32f)

Commit:
fix(Directory): remove erasing print

Fixes #39106

(cherry picked from commit e48fb01d594d62ca1367122d9802177651522f44)

Calinou committed:
Tweak the DirectInput initialization failure message

This closes #36662.

(cherry picked from commit fe09b4d2b39d3ee351f88c604fcdda4b053475ab)

Commit:
SCons: use `OrderedDict` to ensure insertion order of modules

The insertion order for dictionaries is only a language feature for
Python 3.6/3.7+ implementations, and not prior to that.

This ensures that the engine won't be rebuilt if the order of detected
modules changes in any way, as the `OrderedDict` should guarantee
inerstion order.

akien-mga committed:
Merge pull request #39110 from volzhs/godotpayment-consume

Add missing consumePurchase plugin method for GodotPayment

akien-mga committed:
Merge pull request #39121 from Xrayez/py-modules-order

[3.2] SCons: use `OrderedDict` to ensure insertion order of modules

Calinou committed:
Disable antialiasing on the DynamicFont outline as well when requested

This partially addresses
https://github.com/godotengine/godot-proposals/issues/943.

(cherry picked from commit 2919fc73174a3c751405ce46ceed3ef71ea62966)

Commit:
SCons: Allow to read `custom_modules` option via a file

The `custom_modules` option was only read via the command line
by fetching `ARGUMENTS` dictionary directly.

Instead, the option's value can now be read via any existing
configuration files (`custom.py`) as well as command line, while also
updating the environment.

(cherry picked from commit 6304d16915e7dec0b699c2f8789d4bbfef5da66d)

akien-mga committed:
Merge pull request #39032 from mrushyendra/issue-38702

Fix usage req. in Viewport so that ViewportTexture can be sampled as HDR

madmiraal committed:
Do not probe joypads if DirectInput cannot be initialized.

(cherry picked from commit 97a529b1d817aff085777d1bc25272fad60b60b2)

akien-mga committed:
Merge pull request #39086 from madmiraal/fix-39059-3.2

[3.2] Correct Bullet's default Area angular damp value.

madmiraal committed:
Add dinput nullptr checks.

(cherry picked from commit bc49d3412349ed11c03618c96c4889c5620fe5f0)

dreamsComeTrue committed:
Grabbing focus on ProjectList after clicking an item.

Fixes: #39012
(cherry picked from commit 1ec8f59397c157e9f826e1cd53e7790315c969d7)

dreamsComeTrue committed:
Preserving "Sync Scene Changes' and "Sync Script Changes" with Project Settings

Fixes: #38739
(cherry picked from commit 72501ae104d36f5a957ed9ba2ab8f210fb3bf95f)

pihdastudios committed:
Fixes building mono release templates

(cherry picked from commit a77f633161e5f105a89afbba81056b87319c1aad)

lawnjelly committed:
Light2D shadow mask construction fix

When using the default setting (layer 1 set only) nothing is stored in the tscn file for a Light2D, hence it relies on the value in the constructor.

The problem is the constructed value is 1 in Light2D, and -1 in RasterizerCanvas::Light. -1 results in all bits being set so all occluders are shown, rather than just those in layer 1.

This PR changes Rasterizer::Canvas constructor to set to 1. An alternative is to have -1 as the value for layer 1 throughout.

Calinou committed:
Implement zooming using Ctrl + Mouse wheel in the TileMap editor

This was previously implemented in the GridMap editor. This makes
the same feature available in the TileMap editor.

(cherry picked from commit def2059d67b5ec1335f8654e80a2bc19b7faa299)

Calinou committed:
Allow mouse zooming without hovering the texture in the TileSet editor

This partially addresses
https://github.com/godotengine/godot-proposals/issues/968.

(cherry picked from commit d10a5b2928175b313baf709b480620d755983f36)

lawnjelly committed:
Make Light2D respect CanvasLayer layer_min and layer_max

As the masked light list takes no account of layer_min and layer_max, the canvas_layed_id is passed to the _light_mask_canvas_items function where it can be used to reject lights outside the layer range.

madmiraal committed:
Update OSX GUID to SDL uid conversion.

Co-authored-by: John Wakley <johncwakley@users.noreply.github.com>
(cherry picked from commit d04cbbd76b50af563ae6317ca30ba038cc5fc03f)

Calinou committed:
Print errors when calling MIDI input methods on unsupported platforms

This partially addresses #32065.

(cherry picked from commit 9c0d21477617f8bbf771c26787ca5e13d4b29a8a)

Commit:
Prevent crash attaching a script with no languages registered

(cherry picked from commit 52f3cfca6f06b2bb8364049eda345366915c60b8)

Calinou committed:
Add a getter and property for the editor distraction-free mode

(cherry picked from commit cc1859efed2da6587fa4bf3f567167a3c5f20fb6)

Calinou committed:
Add performance hints to the DirectionalLight shadow mode property hint

This also clarifies some parts in the DirectionalLight documentation.

(cherry picked from commit b46756c723ebf7345327f2819d0d8b51bc9d4e55)

EricEzaM committed:
Improved warning shown when autoload cannot be added.

(cherry picked from commit 58caa2c453bed12c9249e7696406775dfa135a9e)

pouleyKetchoupp committed:
Fix inconsistent error messages with Android custom export templates

(cherry picked from commit be22a1f9bcf70f74ab7005bf1e58ff1c877d0cbc)

ntfshard committed:
Fix for Expression class: inner string can be single quoted

(cherry picked from commit ec164b2301034b3aa0fee4e4b93170404a2e4d98)

mrushyendra committed:
Clarify usage of AnimationPlayer with AnimationTree and fill in empty method descriptions

Add clarification to docs to explain that when an `AnimationPlayer`
object is paired with an `AnimationTree`, several properties and
methods exposed by the `AnimationPlayer` class may not work as expected.
The issues occur because an `AnimationTree` runs independently, and
uses its member `AnimationPlayer* player` primarily for its
`Map<StringName, AnimationData> animation_set`.

Added descriptions for `root_motion_track` and `get_root_motion_transform`.

(cherry picked from commit 0451248ff1e878e1b19a65f5db99beaf85645fb8)

akien-mga committed:
Merge pull request #39162 from lawnjelly/light_shadow_mask

3.2 Light2D shadow mask construction fix

bruvzg committed:
[3.2] Fix macOS global menu removal and preserve order.

akien-mga committed:
Merge pull request #39243 from bruvzg/mac_menu_3

[3.2] Fix macOS global menu removal and preserve order.

lawnjelly committed:
GLES2 batching - Add UV precision adjustment for tilemaps

Scaling tilemaps can cause border artifacts around the edges of tiles. This has been traced to precision issues in the GPU. This PR adds an adjustment to allow a minor contraction of the UVs of rects in order to compensate for the incorrect classification of texels across the UV border.

mrushyendra committed:
Enable zero padding with float specifier for format strings

Godot currently supports zero padding for integers, octals and
hexadecimals when using format strings, but not for floats.

This commit adds support for zero padding for floats, thus ensuring
consistent behavior for all types, and making Godot's format specifiers'
behavior closer to c's `printf()`.

Before: `print("<%07.2f>" % -0.2345)` prints `<  -0.23>`.
Now: `print("<%07.2f>" % -0.2345)` prints `<-000.23>`.
     `print("<%7.2f>" % -0.2345)` prints `<  -0.23>`.
(cherry picked from commit b7d835d9cae8625fe32fa837bae6f3e7843f50e7)

vnen committed:
GDScript: Allow get_script() to be considered a type

aaronfranke committed:
Change OK text in snap dialog to OK

(cherry picked from commit 8f8585822401ce5f216976afb4d2f4d05890b9ac)

ThakeeNathees committed:
Fix `is` operation fail on get_script()

Fix: #39244

akien-mga committed:
Merge pull request #39260 from vnen/get_script_is_type

Allow get_script() to be considered a type

theoway committed:
Fixes the right and center alignment bug of rich text label

(cherry picked from commit b2a32d007d125342efb7a87defcac381afaaa70d)

Chaosus committed:
[3.2] Fix shader's length() function parsing in expressions

RandomShaper committed:
Fix excessive bottom cropping in atlas generation

(cherry picked from commit fb7ae73045a0b456c971c26c3c0aa1203aa9a9d7)

akien-mga committed:
Merge pull request #39265 from Chaosus/fix_length_3.2

[3.2] Fix shader's length() function parsing in expressions

ThakeeNathees committed:
shodowed var warning for `for` loop counter

Fix: #39268
(cherry picked from commit 54835a530210c9620973a63dd30ff1d6bbe43a96)

ThakeeNathees committed:
predefined var check for `for` loop counter

(cherry picked from commit e153772de2a52795aefb2f0c11c69396d04c61d5)

akien-mga committed:
Update AUTHORS and DONORS list

New contributor added to AUTHORS:
@jbuck3

Thanks to all contributors and donors for making Godot possible!

(cherry picked from commit 910ddd28171e43649d74e2979c3ed0d68a24dc51)

mbrlabs committed:
Check if the specified Android release keystore exists

(cherry picked from commit 21263a65a9a44fc06ed5311fc600bd72b7f5243f)

akien-mga committed:
Merge pull request #39280 from akien-mga/3.2-cherrypicks

Cherry-picks for the 3.2 branch (future 3.2.2) - 7th batch

mbrlabs committed:
Enable text selection in execute output of EditorNode

(cherry picked from commit 85175c67b858af5cac21aacdfbca78d67ea15120)

akien-mga committed:
Sync controller mappings DB with SDL2 community repo

Synced with gabomdq/SDL_GameControllerDB@6191f6333bd66644c6a208fabb9fd5dabf0a5d43.

(cherry picked from commit 938334cc9eac11faa1df906b0cfcf81911af6112)

mbrlabs committed:
Set "shader_param/" prefix in Shader::has_param()

(cherry picked from commit ffbb211bbe5d4c7cad881d9c34bccf0978226d1f)

Faless committed:
Swtich HTML5 key detection from keyCode to code.

The value of this, does not include the layout.
The code has extra logic to map the unicode value to our keylist,
supporting ASCII and Latin-1.

(cherry picked from commit 0a35b97b62dc1f275a8c84397fe8319ecfc233e7
with help of https://github.com/Faless/godot/tree/js/keyCode)

Faless committed:
Fix WebRTCPeerConnection set_local_description doc

ice_candidate_created should be emitted after set_local_description no
matter the type of the description (assuming no error is returned of
course).

(cherry picked from commit 39bcbf5690032f1f50712c5a908f0e51260471b0)

Calinou committed:
Tweak the GDScript error message about passed argument type mismatch

This makes it less confusing.

This closes https://github.com/godotengine/godot-proposals/issues/670.

(cherry picked from commit 30053de1824c932a6b9f3620b46f8ba056a43c2c)

m4gr3d committed:
Add support for focus awareness

KoBeWi committed:
Allow duplicating files when holding Control

(cherry picked from commit b426d11d8617daa79ccc5fa8c644abb8c8dfecd0)

ThakeeNathees committed:
Debugger stepping to incorrect line fix

Fix: #39296
(cherry picked from commit fc89c5c76d9148f2b3a6789f015087ed3811413c)

ThakeeNathees committed:
Fix: Ctrl + Click not working for subclasses

(cherry picked from commit 95b0a00d2cd282a50c2d7f3a638b7ccfd8648947)

akien-mga committed:
Merge pull request #39305 from m4gr3d/enable_focus_awareness_3_2

[3.2] Add support for focus awareness

KoBeWi committed:
Make sure built-in script warning fits in dialog

akien-mga committed:
Merge pull request #39319 from KoBeWi/this_should_do

Make sure built-in script warning fits in dialog

dominiks committed:
Add note about automatic window title to FileDialog documentation.

(cherry picked from commit d2250a2320abba38a71dd556e32ccaabeb5fcfc7)

KoBeWi committed:
Restore capture caches when missing

(cherry picked from commit 642d91381a0f9077313c76368a1231bf7233c0ab)

Faless committed:
Crypto as a custom instance class.

(cherry picked from commit 87cc283b8b74f78ce58d8510558ff0a8a9e35482)

Faless committed:
Fix editor crash when mbedtls is disabled.

(cherry picked from commit 054f52364fe3c204ca65fca905a5bb3261315e47)

Cevantime committed:
Make text deletion methods public for LineEdit

(cherry picked from commit 2bb44ea5bd57ddaeb7ede96f13736767e4b30342)

RandomShaper committed:
Add shader time scaling

Shaders' `TIME` will be affected by the new shader time scale, that is set via the also new `VisualServer::set_time_scale()`.

giulianob committed:
Tree: Calling update in _gui_input less frequently

(cherry picked from commit d443a13244093a3b094dc52b197f9abb363bdfee)

lostptr committed:
Added a "title" attribute for the link tag in the docs xml

(cherry picked from commit 18c08f65d6b3dd7918a1f6c1d08f30262497af56)

akien-mga committed:
Merge pull request #38995 from RandomShaper/shader_time_scale_3.2

Add shader time scaling (3.2)

akien-mga committed:
Merge pull request #39068 from lawnjelly/kessel_settings_rename

Rename batching project settings in preparation for GLES3

MarcusElg committed:
Fix fbx import assimp error

(cherry picked from commit 29abbccc5f96a53ef74dc9154b60ecc043c33d74)

Calinou committed:
Add an editor tooltip to document gizmo visibility options

See discussion in
https://github.com/godotengine/godot-proposals/issues/716.

(cherry picked from commit 138a4eecb305016a686c43409c554331fcf7c3c7)

mbrlabs committed:
Re-apply basis of newly added node (via drag-n-drop) in editor viewport

aaronfranke committed:
Change the default editor camera rotation to position it in +X +Y +Z

(cherry picked from commit 6c2df6792b59fa6b84f99239797d1d54c8ded764)

Anutrix committed:
Removed variables and #include in EditorSceneImporterAssimp::import_scene that became unused after the recent commit ec1bf96(#39363).

(cherry picked from commit 365c35f30ea3b4749c8a0ca07db8464f9b6bc41c)

Calinou committed:
Document enlarging Particles visibility AABB when they cast shadows

This closes https://github.com/godotengine/godot/issues/17267.

akien-mga committed:
Merge pull request #39376 from Calinou/doc-particles-shadows-aabb

Document enlarging Particles visibility AABB when they cast shadows

akien-mga committed:
Merge pull request #39369 from mbrlabs/basis-fix

[3.2] Re-apply basis of newly added node (via drag-n-drop) in editor viewport

akien-mga committed:
GDScript LSP: Fix crash in notify_client

`latest_client_id` now defaults to `-1` (invalid ID) instead of `0`.

Also fix typo in notification `gdscrip_client/changeWorkspace`,
and fix argument names in method binds.

Fixes #39375.

(cherry picked from commit e34f33711b172a9fe6e5c562dee693ed467576be)

mrushyendra committed:
Fix signal duplication bug when duplicating node with instanced children

Change error checking in `duplicate_signals()` to check for path to
`p_original`, thus adhering to the method used in `duplicate`, instead
of checking for ownership.

akien-mga committed:
VariantParser: Fix crash on malformed vectors

Each time `r_err_str` is set, we should return a parse error.

Removed redundant `return OK;` which were already handled after the big
`if`/`else if`/`else` for `TK_IDENTIFIER`.

Part of #17372.

(cherry picked from commit e7ebda975a4c9f9b8136a571df3c523931358f5b)

akien-mga committed:
PackedScene: Prevent crash when root node has `parent` attribute

The crash happens further down when setting an invalid owner in
`Node::_set_owner_nocheck` but I couldn't figure out how to fix it.

But here the proper fix is to catch the invalid scene file early on
and fail loading it.

Part of #17372.

(cherry picked from commit c080ec5da220474a80789afa33cc4f5612cddb50)

akien-mga committed:
Merge pull request #39134 from abustin/fbx_mesh_compression_fix

Respect 'mesh compression' editor import option in Assimp and glTF importers

YeldhamDev committed:
Add generic file icon and its modulation to the 'FileDialog'

(cherry picked from commit 637927f8035bbffc4ffb0a71115071eaa1a82f3b)

akien-mga committed:
Core: Add hints to run with --verbose when leaking nodes/resources at exit

(cherry picked from commit 2b5545270a60a1a60f57c91bb565131dc03de74d)

madmiraal committed:
Test collision mask before creating constraint pair in Godot physics
broadphase 2D and 3D.

Calinou committed:
Tweak the invalid preset error message to mention `export_presets.cfg`

(cherry picked from commit a7b2f3d41aa47d64ad783d5afb35d08e706bf8ea)

akien-mga committed:
AStar: Make get_closest_point() deterministic for equidistant points

Closes godotengine/godot-docs#3667.
Supersedes #39405.

(cherry picked from commit 187ba4c5a884aaecd97febcdfaaa76466820be07)

madmiraal committed:
Update Rigidbody 2D and 3D sleep documentation.

(cherry picked from commit 1b738a77c15cf3a462ffc8bc509f511d994fc488)

akien-mga committed:
doc: Make File store/get integer methods clearer

Add an example on how to store signed integers on less than 64 bits,
using one bit for the signedness.

(cherry picked from commit cd25d184a5f547d7e7b6332a255908aa00f4ddc7)

akien-mga committed:
Merge pull request #39400 from madmiraal/fix-39374-3.2

[3.2] Test collision mask before creating constraint pair in Godot physics broadphase 2D and 3D.

Commit:
Add description to TileSet.is_tile_bound() method

(cherry picked from commit 00398abf77d464a65f4424b1e6bfda5fdffbe228)

Phischermen committed:
Fix crash when creating new text file with no name

(cherry picked from commit cdb29447b4da77bebc7adc944fe10ef7ea580ebf)

YeldhamDev committed:
Fix scene tree showing up when the root selection is present

(cherry picked from commit f30e4dbf54397646a81b45ab64978e22d4a5c8d9)

Rubonnek committed:
Fixed mismatched disable half float property string

hcmlax committed:
adding check for syskeydown on control and alt

(cherry picked from commit d191a8b9f806641ba5ebdbfe059068dd15056124)

Calinou committed:
Mention ordering caveats for `Dictionary.hash()`

See https://github.com/godotengine/godot/issues/27615.

(cherry picked from commit 77b89263dd37299c3034327a99374d114b42706d)

Calinou committed:
Improve the 3D light documentations

See https://github.com/godotengine/godot-docs/issues/3670.

(cherry picked from commit f7e21d8e8810fa0f0fbaad43e48db31a89412ead)

Calinou committed:
Document the requirement to update the class reference when contributing

Documenting new additions has been an informal requirement for
a few months now. This makes it official.

(cherry picked from commit 6d0e8f9fb70ff00088185f59a627c8f65f5d78ed)

akien-mga committed:
Merge pull request #39426 from Rubonnek/unmatched-half-float

Fixed mismatched disable half float property string

mrushyendra committed:
Account for file deletion and renaming in Export Presets

Ensure that presets are updated with the latest files when
starting up or opening the Project Export dialog. Fixes the
error where Godot would attempt to export deleted files that
were previously selected.

(cherry picked from commit 44094b082d56daa2eab5a8f6c6d73f86d8b18d8b)

akien-mga committed:
Merge pull request #39169 from lawnjelly/light2d_layer_fix

3.2 Make Light2D respect CanvasLayer layer_min and layer_max

akien-mga committed:
Merge pull request #39256 from lawnjelly/tilemap_pixel2

GLES2 batching - Add UV precision adjustment for tilemaps

akien-mga committed:
Merge pull request #39437 from akien-mga/3.2-cherrypicks

Cherry-picks for the 3.2 branch (future 3.2.2) - 8th batch

akien-mga committed:
Merge pull request #38507 from RandomShaper/fix_cleanup_info_3.2

Fix potential crash when listing leaked objects (3.2)

aaronfranke committed:
[3.2] Add GLTF light import

Co-authored-by: K.S. Ernest (iFire) Lee <ernest.lee@chibifire.com>

victoriaderidder committed:
Adds full description for AudioEffectRecord.xml

(cherry picked from commit 718221a2ffd75765d875623a4d3064e1e9120aae)

mrushyendra committed:
Enable Unicode support for RegEx class

Build PCRE2 thirdparty library with unicode support. RegEx objects
in Godot can now be used to recognize unicode strings.

(cherry picked from commit 9f10f6779cfdc15c1304c017bac5ad25de2383b6)

akien-mga committed:
Merge pull request #39448 from mrushyendra/duplicate_signals

[3.2] Fix signal duplication bug when duplicating node with instanced children

akien-mga committed:
doc: Sync classref with current source

(cherry picked from commit fef49bff0ba219af9bdac903cbb3b6953cbc7b8b)

akien-mga committed:
i18n: Sync translations with Weblate

akien-mga committed:
Set version to 3.2.2-rc once again

akien-mga committed:
String: Use ABS macro in padding code

Follow-up to #39261.

(cherry picked from commit cbcc0eacd5cd7af661ec99dcdc8e5a69a8a7ca5d)

Calinou committed:
Document the `in` operator in String, Array and Dictionary classes

This also clarifies the `in` operator behavior in Object.

(cherry picked from commit 7f01f68226c3aa32a372afba3b254402a9448830)

akien-mga committed:
SCons: Fix python2 compatibility after #37198 and #37248

KoBeWi committed:
Fix custom types in node list search

(cherry picked from commit 22f0de8a838486ee38fe9de3d95768ee5ea4fd25)

akien-mga committed:
Merge pull request #39463 from akien-mga/3.2-scons-py2-fixup

[3.2] SCons: Fix python2 compatibility after #37198 and #37248

akien-mga committed:
Merge pull request #39428 from fire/gltf-lights

[3.2] Add GLTF light import

aaronfranke committed:
Simplify some code in platform/uwp/export

(cherry picked from commit 068054002dc43d76d7073b06fe06d63bf9b1a437)

mrushyendra committed:
Fix whole word search slowdown in editor

Reduce repeated iteration through the full text
when counting the number of occurrences of whole
words while searching a file in the editor.

(cherry picked from commit 2433287871ec5ccdf1f85e3e9c057a1a589d3848)

bruvzg committed:
[3.2] Add keyboard layout enumeration / set / get functions (macOS, Windows, Linux/X11).

AndreaCatania committed:
- Added more euler rotation orders support.
- Fixed floating point issue on the old one.
- Fixed the equation on the get_euler_yxz function.
- Added unit tests.

This work has been kindly sponsored by IMVU.

(cherry picked from commit 233130098941c384240a0b7109890e0222f96735)

theoway committed:
Fixes the stopping of animation effects in bbcode after appending

(cherry picked from commit 88b30a29bf6f5497b1f42e33c6dfbff795fd5585)

Calinou committed:
Clarify Node virtual methdods not being called if node is orphan

This closes #39489.

(cherry picked from commit b2ba78ff0f7e83e2b62f3bd0db1afb282781c2ed)

hilfazer committed:
Fixed "Favourite Project" button in Project Manager

madmiraal committed:
Ensure pair callback data is set to null when it's null.

akien-mga committed:
Merge pull request #39504 from madmiraal/fix-39500-3.2

[3.2] Ensure pair callback data is set to null when it's null.

Calinou committed:
Display freelook speed and zoom in units in the 3D editor viewport

Now that the 3D editor grid is infinite and adjusts its step automatically,
this helps people get a better sense of scale when moving around in the
3D viewport.

This also fixes the indicator bar drawing on hiDPI displays.

(cherry picked from commit eaae50fe4ae4585fa5af8ef63c2862153ceb29f5)

skyace65 committed:
Fix label percent visible doc description

(cherry picked from commit a09aeefa1d0b99790d6f5ca964ffd6afd555e8fa)

31 committed:
Fix Image.create_from_data use_mipmaps doc

The Image.create_from_data doc describes the "use_mipmaps" argument as a
way to generate mipmaps, but this method only allocates and loads data.
This can cause confusion, where this function reads more or less data
than expected. Update the doc to be more specific that create_from_data
is loading the mipmaps from the raw data.

(cherry picked from commit 09b324b5b6b956efb2b4a60bd73916bbaf6ff10f)

noidexe committed:
Fix TabContainer emitting spurious tab_selected signals when a theme is changed

- The repaint code was moved outside of set_current_tab() and to a "_repaint()" private function
- _on_theme_changed() will now only call _repaint() and update()
This means _on_theme_changed() will do only what it needs (repainting the TabContainer to account for the new theme)

fixes #39498

(cherry picked from commit 5123006f06e235637b8b1d9faca92eecab476eee)

31 committed:
Fix Material.render_priority doc: no opaque sort

Opaque objects are rendered using the depth buffer, so the end result
appears sorted, but the opaque objects themselves are not sorted.

(cherry picked from commit 608bc8f3dc91930a7d1fc853e6fd8059434de973)

KoBeWi committed:
Re-enable scroll follow on RichTextLabel clear

(cherry picked from commit 46fd51056a1e4a59c9e5b351c12f51cc062db009)

lawnjelly committed:
GLES2 Batching - Fix try_join_item logic for lights

The old logic was incorrect, the first item with lights would prevent joining the next item in case it didn't have lights. Now the check is deferred so that items without lights check to see if the previous item had lights, and if so they prevent a join.

akien-mga committed:
Merge pull request #39531 from lawnjelly/kessel_light_join_next

GLES2 Batching - Fix try_join_item logic for lights

Faless committed:
Use 2-phase setup in JavaScript.

Make the first iteration as soon as the canvas is setup, avoiding a
black screen before the next animation frame is requested.
Also create OS and do setup before syncing FS to avoid crash when IDBFS
access is denied.

Calinou committed:
Document how to resize a SubViewportContainer to avoid stretching

This closes #27534.

(cherry picked from commit f818a99d0b290319665f0cfb4f7024b451fd78aa)

akien-mga committed:
Merge pull request #39538 from Faless/js/2phase

[3.2] Use 2-phase setup in JavaScript.

KoBeWi committed:
Delete bus effect with Delete key

(cherry picked from commit 05656f2b29fbbaeac44346f5c2a67fb926c3e2c6)

mrushyendra committed:
Fix RegEx example in class doc and correct typo

Fix minor errors to do with `search_all` example in RegEx and
typo in RegExMatch class docs.

(cherry picked from commit e5d93f7a66ce93bc8d9db37736aa0e6050ff1172)

mrushyendra committed:
Fix match count for whole word search in editor

Check if a match borders a new line char when incrementing match counts.

(cherry picked from commit 91bdc77d47eba2a0c74c07eb57ce5ee14538b535)

akien-mga committed:
Merge pull request #39501 from hilfazer/favourite-project-button

Fixed 'Favourite' button in Project Manager

akien-mga committed:
GDScript LSP: Fix wrong error checks added in #39385

Reverts `latest_client_id` back to 0, as I misunderstood how the client
IDs are assigned and, without further testing and debugging, I can't
say if this was a bug or a valid default value.
Similarly, a `latest_client_id` of -1 is no longer raising an error.

Fixes #39548.

(cherry picked from commit 786f4ada35a491fb75cdb14f2f34085e4adb8668)

akien-mga committed:
Merge pull request #39502 from bruvzg/kbd_layouts_3

[3.2] Add keyboard layout enumeration / set / get functions.

RandomShaper committed:
Fix wrong internal format for half-float textures

akien-mga committed:
DDS: Fix loading files without DDSD_CAPS or DDSD_PIXELFORMAT

MSDN says:

> When you write .dds files, you should set the DDSD_CAPS and
> DDSD_PIXELFORMAT flags, and for mipmapped textures you should also
> set the DDSD_MIPMAPCOUNT flag. However, when you read a .dds file,
> you should not rely on the DDSD_CAPS, DDSD_PIXELFORMAT, and
> DDSD_MIPMAPCOUNT flags being set because some writers of such a file
> might not set these flags.

https://docs.microsoft.com/en-us/windows/win32/direct3ddds/dds-header

¯\_(ツ)_/¯

Fixes #39516.

(cherry picked from commit 0f21249a2d0f0d6bbd01ef7408a1a9cd0b683793)

Voklen committed:
Add description for "absorbent" and "rough"

(cherry picked from commit 8acdbaa4f27e3ce013995e9b6e2577326f1c7669)

akien-mga committed:
i18n: Sync translations with Weblate

Remove 'de_CH' translation (4% complete) to default to the 100%
complete 'de' one.

Fixes #32625.

Ranoller committed:
(3.2) Fix bad position in Camera2D offset and camera rotation if smothing enabled

Fix https://github.com/godotengine/godot/issues/16323 in 3.2 branch. Apply same solution of https://github.com/godotengine/godot/issues/2074 in rotation and offset.

madmiraal committed:
Add Code::Blocks files to .gitignore.

(cherry picked from commit 990ecf8fc40ad409399f63fe675969b15a41ca0a)

m4gr3d committed:
Clean up the `GodotPlugin` public API.

akien-mga committed:
Merge pull request #39559 from RandomShaper/fix_half_float_textures_3.2

Fix wrong internal format for half-float textures

Chaosus committed:
Use path instead classname to prevent errors for exported visual shaders

akien-mga committed:
Merge pull request #39568 from Ranoller/patch-2

(3.2) Fix bad position in Camera2D offset and camera rotation if smot…

akien-mga committed:
Merge pull request #39571 from Chaosus/fix_vs

[3.2] Use path instead classname to prevent errors for exported visual shaders

madmiraal committed:
Normalise p_up_direction vector in move_and_slide() and
move_and_slide_with_snap() and fix tolerance in
move_and_slide_with_snap() max floor angle.

Calinou committed:
Document Mono-specific classes

This fully documents all Mono-specific classes.

(cherry picked from commit 24a8d432720d57c1cd9bcb89516ea354ea4d459e)

Calinou committed:
Add a suffix to the window title when running from a debug build

Since projects started from the editor or exported in debug mode
run slower than those exported in release mode, this should be
clearly presented to the user.

This partially addresses #20219.

(cherry picked from commit a9c60007a9f7cacd690c09b7f74a5e6f190c5cfb)

akien-mga committed:
Merge pull request #39590 from madmiraal/fix-35744-3.2

[3.2] Normalise p_up_direction vector in move_and_slide()

mrushyendra committed:
Fix code editor Replace tool to work with 'Selection Only' option

Ensures that when the `Selection Only` option is selected, the Replace
tool replaces occurrences of the search phrase only within the selected
region, and in order of occurrence.

(cherry picked from commit 2be3a33cc3e7de4309e6448d79cab2b87fb93e75)

akien-mga committed:
Merge pull request #39263 from ThakeeNathees/is-operation-fail

[3.2] Fix `is` operation fail on get_script()

akien-mga committed:
Merge pull request #39580 from m4gr3d/provide_activity_to_godot_plugin_constructor_3_2

Clean up the GodotPlugin public API.

stijn-h committed:
Search result highlights follow font size

(cherry picked from commit a86521d0bdedc988d7c5a2892dc4d5cf7f01c793)

aaronfranke committed:
Simplify Transform2D get_rotation

(cherry picked from commit 0fa478a050a94da2468f3c3b2216bacd7dedf0ab)

akien-mga committed:
DocData: Skip language-specific ClassDoc without methods/constants

Removes the useless `@C#`, `@NativeScript` and `@VisualScript` entries.

(cherry picked from commit ba0db95909a232e3f4d694294f077c9460ef00ce)

Calinou committed:
Restore previous mouse position when leaving the editor freelook mode

- Remove the crosshair as it no longer serves a purpose (the cursor will
  now appear where the user "expects" it to).

This closes https://github.com/godotengine/godot-proposals/issues/1076.

(cherry picked from commit 4a542e0007b6288ca328eafe5819e58b92bbad15)

bojidar-bg committed:
Fix a random crash in the TileSet editor

Occurred when selecting polygon collision mode, then drawing a single point, then selecting rectangle mode, and finally trying to select a rectangle.

(cherry picked from commit 7850d0316f1100eb0f6b4229fb76784ea20a3e75)

MichaelBelousov committed:
return boxed long when marshalling a godot int to mono runtime in dynamic contexts

(cherry picked from commit 60fcefba699eb6897a69ca47a962e846cedb04be)

mrushyendra committed:
Fix editor texture preview for certain specific dimensions

Ensures no error is issued when attempting to preview a resource that
may be scaled down to <1 pixel when resizing to fit the thumbnail.

(cherry picked from commit c7a8dc7bb9959f3822480f8fe00146ff7fcb3f30)

rsubtil committed:
Expose disable_render_loop property to GDScript

Sl3dge78 committed:
Fix #20467. The "Anim imported" warning gets displayed properly when working on imported anims.

(cherry picked from commit 95cd74fc4b8950cc3623961078981a4a147aec02)

asmaloney committed:
[3.2][macOS] Command-backspace in line edit

Make command-backspace in line edit work like other macOS applications.

If there is a selection, command-backspace deletes the selection.

If there isn't a selection, command-backspace deletes from the cursor to the beginning of the line edit.

This addresses part of godotengine/godot#23548

dreamsComeTrue committed:
Prevent having spaces in signal's method in Connect Dialog

Fixes: #39647
(cherry picked from commit 79f46da1a32192d20c3dd076f1cb59fabed2265b)

Sl3dge78 committed:
Autofilled path gets set properly on animation subresources

(cherry picked from commit d731f43794d13ff9866561090011ad94655f8a24)

lawnjelly committed:
GLES2 Batching - prevent baking on VERTEX use in a shader

Using the operator += in a shader is classified as an 'assign', and so is classified as a write rather than a read. This means that we need to prevent vertex baking on either a write or read (i.e. on usage), rather than just on reads.

akien-mga committed:
Merge pull request #39669 from lawnjelly/kessel_vertex_write_fix

GLES2 Batching - prevent baking on VERTEX use in a shader

akien-mga committed:
Merge pull request #39644 from Ev1lbl0w/feature/disable-rendering

Expose disable_render_loop property to GDScript

akien-mga committed:
Merge pull request #39658 from asmaloney/3.2-macos-command-backspace

[3.2][macOS] Command-backspace in line edit

akien-mga committed:
Export: Rename 'Windows Universal' to 'UWP'

It's otherwise too easy to be confused between 'Windows Desktop' (Win32)
and 'Windows Universal' (UWP).

(cherry picked from commit 08ef0aab160be00c7a82a840c30427190cafd03f)

asmaloney committed:
[3.2] Ensure FileAccessBuffered structs are properly initialized

Calinou committed:
Scroll horizontally when holding Shift with Scroll to Pan enabled

This closes https://github.com/godotengine/godot-proposals/issues/1077.

(cherry picked from commit 84e1c5e4c319fd2cd2312f37633c96f27d53c329)

Calinou committed:
Replace the AssetLib tab icon with a less confusing design

This closes https://github.com/godotengine/godot-proposals/issues/818.

(cherry picked from commit ebabedc61599a2835dc05b0c9df797359451abe4)

akien-mga committed:
Merge pull request #39683 from asmaloney/3.2-FileAccessBuffered-init

[3.2] Ensure FileAccessBuffered structs are properly initialized

pouleyKetchoupp committed:
Set proper file type attribute for OSX zip export

The missing file type in file attributes was causing the file to lose
executable permissions when unzipped with some softwares.

(cherry picked from commit 4501771fd8743e5019fe313e7e6f890d0d68cbd8)

BastiaanOlij committed:
Fix missing slider joystick handling on Windows

(cherry picked from commit aa01b327d600b7c49e2ca6e351e27c6feb62c069)

Voklen committed:
Fix "Fully Axis-Locked" Freelook Navigation Scheme

Co-authored-by: Aaron Franke <arnfranke@yahoo.com>
(cherry picked from commit 29e901c3420fc823fddebd3f8b885e8a816debe3)

mrushyendra committed:
Fix overwrite of manual changes in export_presets.cfg when export dialog is opened

Ensures that behavior when export_presets.cfg file is edited manually
while editor is running remains identical to before commit 44094b0.

(cherry picked from commit 4a7c19e71902bfa9effbd4e22484c4af0963b1bc)

akien-mga committed:
Move Haiku platform port to external repository

The Haiku port now resides at:
https://github.com/godotengine/godot-haiku-platform

While we're happy to support as many Free and Open Source platforms as we can,
we need to put the main focus on the platforms that we can reasonably maintain,
and for Haiku we have been lacking maintainers ever since the port was first
merged in 2015.

The Haiku code has not been compiling and much less working at least since the
release of Godot 3.0, and while some attempts have been made at fixing things,
it's still not functional today in the `3.2` branch (and much less in `master`,
understandably).

Having it in an external repository should hopefully enable Haiku contributors
to fix issues in their own time, and possibly tag versions compatible with
past Godot releases once they are ready.

(cherry picked from commit efcc508ee5c9cf9a6fac0257da7bb36682855e93)

aaronfranke committed:
Update bool documentation to be more clear

(cherry picked from commit 08577fab145d797b204822829477ff2087ac3c39)

madmiraal committed:
Check if point's index exists before adding it to the list of points
that need to split faces when avoiding creating degenerate faces
while merging CSG faces.

naithar committed:
Variant: Fix potential crash when stringifying deleted Object

Fixes #38597.

Calinou committed:
Document how Vector2, Vector3 and Color behave in a boolean context

See #39731.

(cherry picked from commit 227f12478d2d29fc3884388d973a4de654d6651f)

akien-mga committed:
Merge pull request #39729 from madmiraal/fix-39718-3.2

[3.2] Fix creation of duplicate faces in CSG merge faces.

akien-mga committed:
Merge pull request #37500 from madmiraal/check-motion-before-bullet-sweep-3.2

[3.2] Check for motion in cast_motion() before doing Bullet convexSweepTest().

skyace65 committed:
Fix Joint2D doc

(cherry picked from commit a94ba7be599b72a9cac374bdaecbcc9aaaff8a8b)

akien-mga committed:
Revert "return boxed long when marshalling a godot int to mono runtime in dynamic contexts"

This reverts commit 42a31501a8bd858d5560b389712a7c230a4c2074.

It breaks compat as is and needs some further work before it can be
cherry-picked for 3.2.x.

akien-mga committed:
i18n: Sync translations with Weblate

timoschwarzer committed:
Remove GodotPayment android plugin

This is now available in a separate repository
at https://github.com/godotengine/godot-google-play-billing

akien-mga committed:
Merge pull request #39736 from naithar/fix/deleted-object-stringify

Fix for a crash on printing freed object

pouleyKetchoupp committed:
Fix crashes with invalid shapes in BulletPhysicsDirectSpaceState queries

(cherry picked from commit 8b0232bf295c924c157b30f78c9b991fb211dd73)

akien-mga committed:
Revert "Implement Skew in Node2D"

This reverts commit 7f6171018367199c7ac99b7d101572ef5bcd27ba.

See #38868, in its current implementation a small skew value might end up
serialized to scene files due to floating point precision errors, which is
detrimental to VCS.

This can be cherry-picked anew once a fix for #38868 has been found.

bruvzg committed:
macOS, prevent multiple CGDisplayHideCursor calls unpaired with CGDisplayShowCursor.

(cherry picked from commit 7a250b579fb2a95bc1e11eb165b31662efbb4699)

aaronfranke committed:
Avoid errors when the editor camera is inside the focused object

Prevent the gizmo distance from being zero by MAX-ing it with CMP_EPSILON

(cherry picked from commit ed2bbdbbefa190a707f4e09d9bd9c04b564f32ae)

bruvzg committed:
[Windows] Prevent overwriting old cursor handle on multiple mouse_mode changes, ensure mouse_mode is set before `cursor_set_shape` is called to restore cursor.

(cherry picked from commit fc1d1c3557de3339d18df2639e12edca33a774a5)

bruvzg committed:
Fix use uninitialized `ofs` variable introduced in 39701.

(cherry picked from commit 439c97e0ff7b33fb1fc4e63353fb046efd8476b8)

akien-mga committed:
Merge pull request #39744 from timoschwarzer/remove-godot-payment-plugin

Remove GodotPayment android plugin (3.2)

akien-mga committed:
doc: Sync classref with current source

ShlomiRex committed:
Working 3.2 with file_dialog need to fix bugs

MCrafterzz commented:
@ShlomiRex you gotta rebase this...

aaronfranke commented:
@ShlomiRex It seems that you rebased incorrectly. Please read this documentation article for more information: https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html

Alternatively, if you want to use a GUI, watch this video: https://www.youtube.com/watch?v=0XKy8YpP5Ww
